src/Npgsql/NpgsqlConnector.cs

```diff
@@ -175,17 +175,15 @@ int ReceiveTimeout
             set
             {
                 // TODO: Socket.ReceiveTimeout doesn't work for async.
-                if (value != _currentTimeout)
-                    _socket.ReceiveTimeout = _currentTimeout = value;
+                Debug.Assert(_socket != null);
+
+                if ((_socket != null) && (value != _socket.ReceiveTimeout))
```

Comment:
Removing _currentTimeout seems like an unrelated and unnecessary change. The purpose of _currentTimeout is to cache the timeout so that we don't have to read _socket.ReceiveTimeout every time - that's a relatively expensive system call.

Reply:
Didn't realize the call was expensive. I needed to set the timeout before _socket was set and not a big fan of requiring mutiple var set without some common method. Maybe since it's a one time case during Connect() it doesn't warrant this concern, but wanted to avoid having to put this setter pair everywhere (in case this is needed somewhere else):

socket.ReceiveTimeout = x;
_currentTimeout = x;

Strange that MS wouldn't cache the value themselves... Wonder why?

Reply:
Also the property is very much tied to the socket. So it should be set on creation/destruction. Maybe the right move here is to create a derived socket class and cache it there.

```
using System.Net.Sockets;

namespace Npgsql
{
    internal class NpgsqlSocket : Socket
    {
        private int currentTimeoutMs;

        public NpgsqlSocket(SocketType socketType, ProtocolType protocolType) : base(socketType, protocolType)
        {
        }

        public NpgsqlSocket(AddressFamily addressFamily, SocketType socketType, ProtocolType protocolType) : base(addressFamily, socketType, protocolType)
        {
        }

        public void SetReceiveTimeout(int timeoutMs)
        {
            if (currentTimeoutMs != timeoutMs)
            {
                base.ReceiveTimeout = currentTimeoutMs = timeoutMs;
            }
        }
    }
}
```
This would also give you flexibility later on if you want to push the read/write buffers into this class.

Reply:
> Strange that MS wouldn't cache the value themselves... Wonder why?

I'm not actually sure that the socket implementation doesn't cache, but it would be odd if .NET had a cache for each and every socket option., especially since reading these is probably not a common or perf-sensitive thing in general. Also, that would all be internal implementation details anyway, and I prefer we cache it ourselves because there's no downside to that.

> Also the property is very much tied to the socket. So it should be set on creation/destruction.

I don't think there's any reason to set the timeout on destruction. Re creation, it seems sufficient to set it before the first actual read occurs after connection, no?

I see your point with NpgsqlSocket, but we basically already do that today - but in a wrapper (NpgsqlReadBuffer) rather than a subclass. That's pretty much the same idea, just with composition instead of inheritance. I don't think we have any plans to merge the buffers and the socket.

In any case, take a look at #3229 which fixes this as part of a larger overhaul of the connection timeout mechanism.

Reply:
Ok, can test against my scenario tomorrow AM. Do you need me to fix this up for a backport or close it out?